it/src/main/java/org/corfudb/universe/universe/docker/DockerUniverse.java

```diff
@@ -0,0 +1,206 @@
+package org.corfudb.universe.universe.docker;
+
+import com.google.common.collect.ImmutableMap;
+import com.spotify.docker.client.DockerClient;
+import com.spotify.docker.client.messages.NetworkConfig;
+import lombok.Builder;
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.universe.group.docker.DockerCorfuCluster;
+import org.corfudb.universe.group.Group;
+import org.corfudb.universe.group.Group.GroupParams;
+import org.corfudb.universe.universe.Universe;
+import org.corfudb.universe.universe.UniverseException;
+import org.corfudb.universe.util.ClassUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Represents Docker implementation of a {@link Universe}.
+ */
+@Slf4j
+public class DockerUniverse implements Universe {
+    /**
+     * Docker parameter --network=host doesn't work in mac machines,
+     * FakeDns is used to solve the issue, it resolves a dns record (which is a node name) to loopback address always.
+     * See Readme.md
+     */
+    private static final FakeDns FAKE_DNS = FakeDns.getInstance().install();
+
+    private final UniverseParams universeParams;
+
+    private final DockerClient docker;
+    private final DockerNetwork network = new DockerNetwork();
+
+    private final ConcurrentMap<String, Group> groups = new ConcurrentHashMap<>();
+    private final String clusterId;
+    private final AtomicBoolean initialized = new AtomicBoolean();
+
+    @Builder
+    public DockerUniverse(UniverseParams universeParams, DockerClient docker) {
+        this.universeParams = universeParams;
+        this.docker = docker;
+        this.clusterId = UUID.randomUUID().toString();
+
+        Runtime.getRuntime().addShutdownHook(new Thread(this::shutdown));
+    }
+
+
+    /**
+     * Deploy a {@link Universe} according to provided parameter, docker client, docker network, and other components.
+     * The instances of this class are immutable. In other word, when the state of an instance is changed a new
+     * immutable instance is provided.
+     *
+     * @return Current instance of a docker {@link Universe} would be returned.
+     * @throws UniverseException this exception will be thrown if deploying a {@link Universe} is not successful
+     */
+    @Override
+    public DockerUniverse deploy() {
+        log.info("Deploying universe: {}", universeParams);
+
+        if (!initialized.get()) {
+            network.setup();
+            initialized.set(true);
+        }
+
+        createAndDeployGroups();
+
+        return this;
+    }
+
+    @Override
+    public void shutdown() {
+        log.info("Shutdown docker universe: {}", clusterId);
```

Comment:
We need to save the data logs and the console logs for each node, so that they can be inspected.

Reply:
Docker provides that functional out of the box,
just execute `docker logs containerName` 
But actually, we just need a centralized logging system like ELK, which is an another question